### ISSUE TITLE
Azure file inline volume mount secret lookup is broken in aks 1.19.7

### ISSUE THREAD

**Issue by tehho** [Azure/AKS#2108]
**What happened**:
Upgrading to 1.19.7 broke azure files secret namespace lookup.

**What you expected to happen**:
When creating a azure files volume i expect the pod to search in its own namespace not default.

**How to reproduce it (as minimally and precisely as possible)**:
```
apiVersion: v1
kind: namespace
metadata:
    name: test
--
apiVersion: v1
kind: Secret
metadata:
    name: test-azure-files
    name: test
data:
    azurestorageaccountkey: <your-key-here>
    azurestorageaccountname: <your-accountname-here>
--
apiVersion: v1
kind: Pod
metadata:
  name: example
  namespace: test
spec:
  containers:
    - name: test
      image: 'ubuntu:latest'
      volumeMounts:
        - mountPath: /test
          name: test
  volumes:
    - azureFile:
        secretName: test-azure-files
        shareName: <your-share-here>
      name: test
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): aks 1.19.7
- Size of cluster (how many worker nodes are in the cluster?) differs per environment.
- General description of workloads in the cluster (e.g. HTTP microservices, Java app, Ruby on Rails, machine learning, etc.) Http microservice. 
- Others:

**Comment by zhiweiv** [comment]
@andyzhangx

**Comment by andyzhangx** [comment]
that is actually a bug fixed in 1.19.7: https://github.com/kubernetes/kubernetes/pull/97417
pls specify `secretNamespace` in azure file PV or azure file storage class if pod wants to specify a different namespace other than `default`, here is an example: https://github.com/andyzhangx/demo/blob/master/pv/pv-azurefile.yaml#L12

**Comment by tehho** [comment]
@andyzhangx We dont use pv or cis.
Following [https://github.com/kubernetes/examples/tree/master/staging/volumes/azure_file](https://github.com/kubernetes/examples/tree/master/staging/volumes/azure_file) we use the "Mount volume directly in Pod" option.
Tried to add secretNamespace to the azureFiles block but got an error saying the field is not supported. 
Is there anyway to do a direct volume on the pod or do we need to switch to pv?

**Comment by andyzhangx** [comment]
> @andyzhangx We dont use pv or cis.
> Following https://github.com/kubernetes/examples/tree/master/staging/volumes/azure_file we use the "Mount volume directly in Pod" option.
> Tried to add secretNamespace to the azureFiles block but got an error saying the field is not supported.
> Is there anyway to do a direct volume on the pod or do we need to switch to pv?

@tehho pod spec(using inline volume) does not have `secretNamespace` field, why do you want to use pod, if node is down, pod would disappear, right? Use deployment + PV/PVC is common solution in production.

**Comment by tehho** [comment]
We use deployments but would like to use it inline on the deployment for least effort on the fix :)

**Comment by pbloigu** [comment]
This very same issue bit me just an hour ago and was a PITA to figure out. Please _at least_ add a line or two in the documentation about this. 

> We use deployments but would like to use it inline on the deployment for least effort on the fix :)

My problem exactly. The volume is directly specified in the deployment w/o any PV/PVC. It used to work, and now suddenly it doesn't.

**Comment by andyzhangx** [comment]
one workaround is create secret in `default` namespace, we are in the process of moving to Azure File CSI driver, in CSI driver, it cannot get the namespace of pod, so anyway it will break at last, if secret is not in `default` namespace, the `secretNamespace` must be specified.

**Comment by pbloigu** [comment]
> one workaround is create secret in `default` namespace, we are in the process of moving to Azure File CSI driver, in CSI driver, it cannot get the namespace of pod, so anyway it will break at last, if secret is not in `default` namespace, the `secretNamespace` must be specified.

Yeah, in the end I ended up moving the secret to the default namespace. My criticism was directed towards changing the current working implementation silently to something that works differently. 

Now, when do you expect this Azure File CSI Driver becomes available? I'm currently considering updating the rest of my clusters and I need to decide between moving the secret to the default namespace vs. moving to PV+PVC vs. waiting for the CSI Driver.

**Comment by norbertkovacs** [comment]
> one workaround is create secret in `default` namespace, we are in the process of moving to Azure File CSI driver, in CSI driver, it cannot get the namespace of pod, so anyway it will break at last, if secret is not in `default` namespace, the `secretNamespace` must be specified.

We can't use default namespace and tried to specify secretNamespace right below secretName value but getting the following error message. 

error: error validating "something.yaml": error validating data: ValidationError(Deployment.spec.template.spec.volumes[0].azureFile): unknown field "secretNamespace" in io.k8s.api.core.v1.AzureFileVolumeSource; if you choose to ignore these errors, turn validation off with --validate=false

**Comment by andyzhangx** [comment]
> > one workaround is create secret in `default` namespace, we are in the process of moving to Azure File CSI driver, in CSI driver, it cannot get the namespace of pod, so anyway it will break at last, if secret is not in `default` namespace, the `secretNamespace` must be specified.
> 
> Yeah, in the end I ended up moving the secret to the default namespace. My criticism was directed towards changing the current working implementation silently to something that works differently.
> 
> Now, when do you expect this Azure File CSI Driver becomes available? I'm currently considering updating the rest of my clusters and I need to decide between moving the secret to the default namespace vs. moving to PV+PVC vs. waiting for the CSI Driver.

@pbloigu I think you could move to PV+PVC with in-tree(default) azure file driver first, and set `azureFile.secretNamespace` in your config, does not need to wait for CSI driver, here is an example:
```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  name: pv-azurefile
spec:
  capacity:
    storage: 5Gi
  accessModes:
    - ReadWriteMany
  azureFile:
    secretName: azure-secret
    secretNamespace: default
    shareName: k8stest
    readOnly: false
```

Inline volume only supports `default` namespace, original behavior(search for pod namespace) is actually an incorrect behavior(fixed by https://github.com/kubernetes/kubernetes/pull/97417), when CSI migration switch is turned on, there is no way to migrate to CSI driver since in CSI driver, it is not aware of pod namespace.

**Comment by tomasr** [comment]
Finding out this issue the hard way has been very disappointing. I can understand that it was a "bug", but it's been around for so long that fixing it now, this way, in a minor update, and with absolutely no heads up, was completely unacceptable.

I've now had to spend multiple hours sorting the mess this created on our system, and completely rewriting chunks of our provisioning infrastructure. I would have been happy to do this properly, if MS had only bothered to communicate costumers of this change.

What a mess.

**Comment by mvvveen** [comment]
Hit the same issue yesterday and found out the hard way too.
Also i didnt make use of pvc/pv for azurefile as the documentation on both ms and k8s states

**Comment by winromulus** [comment]
Hit this problem today the hard way as well. The 1.19.7 version should still be in preview considering this kind of issue.

**Comment by winromulus** [comment]
@andyzhangx 
Getting the following: `error validating data: ValidationError(Deployment.spec.template.spec.volumes[1].azureFile): unknown field "secretNamespace" in io.k8s.api.core.v1.AzureFileVolumeSource`
Is there something we need to update to support this?

**Comment by tehho** [comment]
> @andyzhangx
> Getting the following: `error validating data: ValidationError(Deployment.spec.template.spec.volumes[1].azureFile): unknown field "secretNamespace" in io.k8s.api.core.v1.AzureFileVolumeSource`
> Is there something we need to update to support this?

"secretNamespace" is not supported on the direct volume mount, only with persistent volume mount. 
I have made a PR (https://github.com/kubernetes/kubernetes/pull/98967) to fix this but since this would introduce a API change and direct volume is to be deprecated with the support for CSI it's not likely to be merged

**Comment by winromulus** [comment]
This completely breaks our system. We can't use the default namespace to host that secret since it conflicts and we rely on namespace isolation.

**Comment by amjadaljunaidi** [comment]
@winromulus could you please try to recreate the PV ( delete then create it )

**Comment by winromulus** [comment]
@amjadaljunaidi - We managed to fix this by moving away from directly attached volumes (inline) to PV+PVC but we had to break a lot of other things to work around this issue:
- PVs are cluster-level resources so we had to break the namespace level permissions in order to allow the deployment of the PVs (this defeats the entire purpose of Azure DevOps Environments for Kubernetes where the serviceAccount has access to only one namespace)
- We had to update everything on the fly (templates etc. and workaround third party Helm charts) to make changes to deployments to use PVCs

Personally I'm extremely disappointed in the lack of visibility into this breaking change done by AKS team for a Kubernetes revision bump. If this was in a minor version bump and in preview, I could have understood and I would have expected this. We checked our application for Kubernetes changes that might be affected in the new stable version, not AKS specific hidden gems.

For inline volumes it would have been good to have `secretNamespace` support as well so AKS would support a normal Kubernetes feature instead of adding a limitation. Even if this was still a breaking change, it would have been easier to fix than saying "Do not use Pods, go with Deployments". We might want Pods.

I want to thank the AKS support team that found this ticket for us.

**Comment by amjadaljunaidi** [comment]
@winromulus I totally agree with you and the reason for that changed because there is some modification in next version that could impact your cluster hence we started to that for this version

we will announce this in release notes in the future to avoid any conflict

**Comment by markwooff** [comment]
If inline file mounts are being deprecated, it should be made clear to AKS users since the documentation still shows this as the first way to mount File Shares onto a Pod: https://docs.microsoft.com/en-us/azure/aks/azure-files-volume

Will be looking at converting our inline file mounts to PV + PVC but would have been great to know before upgrading AKS versions.

**Comment by andyzhangx** [comment]
> If inline file mounts are being deprecated, it should be made clear to AKS users since the documentation still shows this as the first way to mount File Shares onto a Pod: https://docs.microsoft.com/en-us/azure/aks/azure-files-volume
> 
> Will be looking at converting our inline file mounts to PV + PVC but would have been great to know before upgrading AKS versions.

@markwooff sure, if you are going to upgrade to k8s 1.18.15, 1.19.7, 1.20.2, 1.21.0, pls change to PV/PVC(specify `secretNamespace` in PV) or move secret to `default` namespace. The inline volume mount is deprecated in CSI driver.

**Comment by palma21** [comment]
Hi folks, sorry for this. We've updated the docs, but more importantly this escaped our net. This was an upstream change in k8s, which impact we missed and so we didn't pre-announce / announce with the introduction of 1.19.7. We try to be diligent with these  and apologize for the big miss.
Even though we always recommend reading the upstream [changelog](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.19.md#v1197) for each version I don't think this one would give it straight away (in the same way that we missed it). We'll post mortem this so it doesn't happen again.

**Comment by joelbernstein** [comment]
Hi, I have this problem too, I think moving to PV is not an option as I am mounting directly into the pod (actually a Job).

If the AzureFile CSI driver would be sufficient, please advise.

Currently I have hacked a workaround using the `default` namespace but this will not be a usable fix. I am shocked this has been released without testing, as this must break things for lots of users. 

PS @andyzhangx I don't think these comments about "well, don't do that then" are helpful. Don't release bugs like this, then, if you don't like the paying users pointing out how it hurts them.

**Comment by joelbernstein** [comment]
BTW with the direct to pod mount, and `secretNamespace: foo` we get a ridiculous error about no such secret as `default/foo/yourSecretName`. It's just ridiculous, your testing strategy must be not fit for purpose.

This needs fixing ASAP, please. And not by telling people to move to a different, unreleased storage driver.

**Comment by poochwashere** [comment]
This bit me as well and caused me to waste a bunch of time before finding this issue. For now I have deployed all of our file share secrets in the `default` namespace so that the pods and jobs can be scheduled.

**Comment by andyzhangx** [comment]
the original issue was reverted in 1.18.18, 1.19.10, 1.20.6, 1.21.0, https://github.com/kubernetes/kubernetes/pull/100563

**Comment by joelbernstein** [comment]
Which means?

**Comment by andyzhangx** [comment]
> the original issue was reverted in 1.18.18, 1.19.10, 1.20.6, 1.21.0, [kubernetes/kubernetes#100563](https://github.com/kubernetes/kubernetes/pull/100563)

this inline volume issue was fixed 1.18.18, 1.19.10, 1.20.6, 1.21.0

**Comment by BojanOro** [comment]
I'm struggling to find documentation on when to expect versions to come to AKS when released from k8s upstream. 

Anybody know when we can expect to see AKS 1.20.6 with this fix?

**Comment by daseth** [comment]
> I'm struggling to find documentation on when to expect versions to come to AKS when released from k8s upstream.
> 
> Anybody know when we can expect to see AKS 1.20.6 with this fix?

Microsoft support has suggested that we should expect a new minor version by the end of this month. Looking at the current date, it seems quite unlikely

**Comment by vinusafeway** [comment]
Hi Team, Can we have ETA or date when we can expect any of the fixed version mentioned

**Comment by djsly** [comment]
Looking for the 1.19.10 release date as well.  We have many users who don't have access to create PV and they don't have access to the default namespace. 

Right now, users had the full control on providing their own secret to access their own SA in their deployment resource using the inline Azure File. 

While 1.19.10 looks like a great revert and will bring back the original support, I'm even more interested in the recommended solution with the CSI Azure File Driver,  how we can ensure that Users still don't get access to the default namespace, and are not given access to create PV.

**Comment by deepakpandiyan** [comment]
@andyzhangx Customer currently done the workaround in  1.19.7 , if he needs to upgrade to1.19.10 do we need follow any procedure before upgrade 
https://docs.microsoft.com/en-us/azure/aks/azure-files-volume#mount-file-share-as-an-persistent-volume

**Comment by elkrammer** [comment]
I've unfortunately just hit this issue too and learning about it the hard way.
In our case we were on `1.18.10` and upgraded to `1.18.17` - a minor upgrade like this should not result in a breakage of this magnitude. This is going to require lots of engineering hours to fix.

Microsoft support has suggested we upgrade to `1.20.x` or `1.21.x` as these versions are not affected by this problem - can anyone confirm?

**Comment by BojanOro** [comment]
> I've unfortunately just hit this issue too and learning about it the hard way.
> In our case we were on `1.18.10` and upgraded to `1.18.17` - a minor upgrade like this should not result in a breakage of this magnitude. This is going to require lots of engineering hours to fix.
> 
> Microsoft support has suggested we upgrade to `1.20.x` or `1.21.x` as these versions are not affected by this problem - can anyone confirm?

We're experiencing issues with 1.20.5 and expecting a fix in 1.20.6. I'd suggest staying where you are for now, no reason to upgrade to another broken version. No word from MS support yet on when to expect the fixed versions.

**Comment by joelbernstein** [comment]
It's not just me, right? This is mad, isn't it? How can this STILL be an
issue? When will it be resolved? It's just embarrassing.

**Comment by andyzhangx** [comment]
there will be fixed versions begin to roll out this week, and user could upgrade to the fixed version

**Comment by djsly** [comment]
k8s 1.19.10 and 1.20.6 support will be rolling out to AKS this week, the upgrade process takes around ~weeks. Once the version is available in your region, you (the aks admin) will be able to perform an AZ AKS UPGRADE command to update your cluster to the latest k8s version that's available.

Hopefully that makes it clearer

**Comment by joelbernstein** [comment]
A lot clearer. Thank you. Not sure how you got that from what Mr Zhang said
but I appreciate the clarification!

**Comment by elkrammer** [comment]
@djsly what about `1.18`?

**Comment by vinusafeway** [comment]
Hi Team,

Could you please confirm

  1.  Availability of 1.19.10 date
  2.  If we are moving with 1.19.10,1.19.7/1.19.9 with PV fix do we need to take care anything else

Thanks
Vinu Kumar
________________________________
From: deepakpandiyan ***@***.***>
Sent: Monday, May 10, 2021 8:50 AM
To: Azure/AKS ***@***.***>
Cc: Vinu Kumar ***@***.***>; Comment ***@***.***>
Subject: EXTERNAL EMAIL: Re: [Azure/AKS] Azure file inline volume mount secret lookup is broken in aks 1.19.7 (#2108)


@andyzhangx<https://clicktime.symantec.com/31a6fyUBUaAiWW29jGrVuh7Vc?u=https%3A%2F%2Fgithub.com%2Fandyzhangx> Customer currently done the workaround in 1.19.7 , if he needs to upgrade to1.19.10 do we need follow any procedure before upgrade
https://docs.microsoft.com/en-us/azure/aks/azure-files-volume#mount-file-share-as-an-persistent-volume<https://clicktime.symantec.com/3SSnMKUS5UzaiBRrYULJnrz7Vc?u=https%3A%2F%2Fdocs.microsoft.com%2Fen-us%2Fazure%2Faks%2Fazure-files-volume%23mount-file-share-as-an-persistent-volume>

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://clicktime.symantec.com/3FbKCGkZTvfQGf5xTY6svKB7Vc?u=https%3A%2F%2Fgithub.com%2FAzure%2FAKS%2Fissues%2F2108%23issuecomment-836111900>, or unsubscribe<https://clicktime.symantec.com/3PBnfw3ccATvtp1JNASDuPh7Vc?u=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAQUOPLCLM5IDMLDPPQPYHPLTM5GI3ANCNFSM4XK7EUVA>.

________________________________
Warning: All e-mail sent to this address will be received by the corporate e-mail system, and is subject to archival and review by someone other than the recipient. This e-mail may contain proprietary information and is intended only for the use of the intended recipient(s). If the reader of this message is not the intended recipient(s), you are notified that you have received this message in error and that any review, dissemination, distribution or copying of this message is strictly prohibited. If you have received this message in error, please notify the sender immediately.
________________________________

**Comment by alftio** [comment]
Update:

[Release 2021-05-13](https://github.com/Azure/AKS/releases/tag/2021-05-13)

Preview Features
The CSI Secret Store AKS Addon is now in Public Preview. See more [here](https://docs.microsoft.com/en-us/azure/aks/csi-secrets-store-driver).
Component Updates
Upgrade azuredisk/azurefile CSI Driver to v1.2.0 (currently in preview).

**Comment by evenh** [comment]
I contacted Azure support and got this answer on May 12th

>Hi Even,
> 
>We understand the struggles this may be causing to our costumers and is being worked on.
> 
>I just got more feedback on this and it’s expected to have this version (1.19.10) fully rolled out until the end of this month (May 2021). Roll out will won’t be done on all regions at once and  is expected to start somewhere next week.
>
>[…]

**Comment by evenh** [comment]
As far as I can see, the updated patch versions is still not available the regions I have checked: `westeurope northeurope westus2 centralus northcentralus eastus`. The latest answer from Azure Support states:
> These releases are still scheduled start until the end of this Month.

**Comment by gopher194** [comment]
Hey, did just hit this surrealist issue while upgrading my deployments ... WTH!
Still need to put the secret in default?

**Comment by gopher194** [comment]
Ok, I got my answer:
`Error: error validating "": error validating data: ValidationError(Deployment.spec.template.spec.volumes[0].azureFile): unknown field "secretNamespace" in io.k8s.api.core.v1.AzureFileVolumeSource`

**Comment by milunko** [comment]
Temporary fix for this solution, until you upgrade to 1.20.x witch does support the inline definition:

kg secret -A | grep YOUR_SECRETS | awk -F ' ' '{print $1, $2}' | xargs -l1 -- sh -c 'kubectl get secret -n $1 $2 -o yaml | sed "s/namespace: $1/namespace: default/" | kubectl apply --namespace default -f -' --

When you upgrade to 1.20.x . remove these secrets.

**Comment by andyzhangx** [comment]
fixed versions(`1.21.1`, `1.20.7`, `1.19.11`, `1.18.19`) are already in `eastus2euap` region, wait for a few more days it would roll out other regions.
```
# az aks get-versions -l eastus2euap -o table
The behavior of this command has been altered by the following extension: aks-preview
KubernetesVersion    Upgrades
-------------------  ------------------------
1.21.1(preview)      None available
1.20.7               1.21.1(preview)
1.20.5               1.20.7, 1.21.1(preview)
1.19.11              1.20.5, 1.20.7
1.19.9               1.19.11, 1.20.5, 1.20.7
1.18.19              1.19.9, 1.19.11
1.18.17              1.18.19, 1.19.9, 1.19.11
```

**Comment by andyzhangx** [comment]
the in-tree driver fix has been rolled out to all regions.
Azure File CSI driver v1.3.0 supports inline volume feature also, could refer to  https://github.com/kubernetes-sigs/azurefile-csi-driver/blob/master/deploy/example/e2e_usage.md#option3-inline-volume

```yaml
  volumes:
    - name: persistent-storage
      csi:
        driver: file.csi.azure.com
        volumeAttributes:
          shareName: EXISTING_SHARE_NAME
          secretName: azure-secret
```
